src/main/java/cubicchunks/server/ServerCubeCache.java

```diff
@@ -269,151 +263,156 @@ public boolean chunkExists(int cubeX, int cubeZ) {
 	//==============================
 
 	@Override
-	public boolean cubeExists(int cubeX, int cubeY, int cubeZ) {
-		long columnAddress = AddressTools.getAddress(cubeX, cubeZ);
-		Column column = this.loadedColumns.get(columnAddress);
-		if (column == null || column.unloaded) {
-			return false;
-		}
-		Cube cube = column.getCube(cubeY);
-		return cube != null && !cube.unloaded;
+	public Cube getCube(int cubeX, int cubeY, int cubeZ) {
+		return getCube(new CubeCoords(cubeX, cubeY, cubeZ));
 	}
 
-	public boolean cubeExists(CubeCoords coords) {
-		return this.cubeExists(coords.getCubeX(), coords.getCubeY(), coords.getCubeZ());
+	@Override
+	public Cube getCube(CubeCoords coords) {
+		return getCube(coords, Requirement.GENERATE);
 	}
 
 	@Override
-	public Column getColumn(int columnX, int columnZ) {
-		Column column = this.loadedColumns.get(AddressTools.getAddress(columnX, columnZ));
-		if(column != null) {
-			column.unloaded = false;
-		}
-		return column;
+	public Cube getLoadedCube(int cubeX, int cubeY, int cubeZ) {
+		return getLoadedCube(new CubeCoords(cubeX, cubeY, cubeZ));
 	}
 
 	@Override
-	public Cube getCube(int cubeX, int cubeY, int cubeZ) {
-		Column column = getColumn(cubeX, cubeZ);
-		if (column == null) {
-			return null;
-		}
-		Cube cube = column.getCube(cubeY);
-		if(cube != null) {
+	public Cube getLoadedCube(CubeCoords coords) {
+		Cube cube = cubemap.get(coords);
+		if(cube != null){
 			cube.unloaded = false;
 		}
 		return cube;
 	}
 
-	public Cube getCube(CubeCoords coords) {
-		return this.getCube(coords.getCubeX(), coords.getCubeY(), coords.getCubeZ());
-	}
+	@Override
+	@Nullable
+	public Cube getCube(CubeCoords coords, Requirement req){
+		if(coords.getMinBlockY() < worldServer.getMinHeight()){
+			throw new IllegalStateException("Why are you getting a Cube so low? got: " + coords);
+		}
 
-	public void loadCube(int cubeX, int cubeY, int cubeZ, LoadType loadType) {
+		Cube cube = getLoadedCube(coords);
+		if(req == Requirement.CACHE || 
```

Comment:
@Boreeas CACHE is a lot like getLoadedCube()... It could be useful in a spectator mode that only sends you cached cubes (so you can fly around and see what is loaded with out loading any chunks your self).